File: schgen/constraints.py
"""Layout constraints from typed ports — JLCPCB JLC04161H-7628 stackup.

Emits, from the PORT types declared in the subsystem models:
  - ``layout_constraints.kicad_dru``  — KiCad custom design rules (conditions
    keyed on net-class names; import the classes from the CSV);
  - ``layout_constraints.csv``        — the human/import table: one row per
    typed port with class, geometry, pairing and length-match group.

Stackup geometry — JLC04161H-7628 (4 layer, 1.6 mm; locked in
carrier/PLAN.md round 2). Outer-layer microstrip referenced to the L2/L3
plane through one sheet of 7628 prepreg (0.2104 mm, er~4.6). The width/gap
numbers below are JLCPCB's own impedance-calculator output for THIS stackup
(community-verified: eevblog thread "JLCPCB Impedance Calculator can't get
100R or 90R"; jlcpcb.com/impedance lists the stackup, the calculator gives
the geometry):

  90R  differential (USB 2.0 HS):   width 10.28 mil = 0.2611 mm, gap 8 mil = 0.2032 mm
  100R differential (TMDS/LVDS/MIPI/MDI): width 8.08 mil = 0.2052 mm, gap 8 mil = 0.2032 mm

NOTE: the often-quoted "90R at 0.127/0.127 mm" geometry belongs to the
THINNER JLC04161H-3313 prepreg (0.0994 mm), not the 7628 stackup this
project locked. Re-run jlcpcb.com/pcb-impedance-calculator before tape-out
if the stackup changes.

Length-match policy (documented defaults, override per-design later):
  usb_hs_pair: intra-pair skew <= 1.27 mm (USB 2.0 HS budget)
  diff_pair:   intra-pair skew <= 1.27 mm (1000BASE-T MDI class)
  tmds_pair:   intra-pair skew <= 0.15 mm; inter-pair <= 5 mm (HDMI class)
  sd_bus:      bus length match <= 2.5 mm to CLK
"""

from __future__ import annotations

import csv
from dataclasses import dataclass
from pathlib import Path

from schgen.model import NetClass

MM_PER_MIL = 0.0254


@dataclass(frozen=True)
class DiffGeometry:
    impedance: int
    width_mm: float
    gap_mm: float
    source: str


# JLC04161H-7628, outer layer vs L2/L3 plane (see module docstring)
GEOMETRY: dict[int, DiffGeometry] = {
    90: DiffGeometry(90, round(10.28 * MM_PER_MIL, 4), round(8 * MM_PER_MIL, 4),
                     "JLCPCB calculator, JLC04161H-7628 outer/L2"),
    100: DiffGeometry(100, round(8.08 * MM_PER_MIL, 4), round(8 * MM_PER_MIL, 4),
                      "JLCPCB calculator, JLC04161H-7628 outer/L2"),
}

INTRA_PAIR_SKEW_MM = {
    "usb_hs_pair": 1.27,
    "diff_pair": 1.27,
    "tmds_pair": 0.15,
}
SD_BUS_MATCH_MM = 2.5


def _net_class(kind: str, impedance: int | None, level_v: float | None) -> str:
    if kind == "usb_hs_pair":
        return "DP90_USB"
    if kind == "tmds_pair":
        return "DP100_TMDS"
    if kind == "diff_pair":
        return f"DP{impedance}_DIFF"
    if kind == "i2c":
        return "I2C"
    if kind == "sd_bus":
        lv = f"{level_v:g}".replace(".", "V")
        return f"SD_{lv}"
    return "Default"


def export(sheets, outdir: Path) -> tuple[Path, Path]:
    """``sheets``: list of schgen.link.SheetCircuit. Writes the .kicad_dru
    and .csv into ``outdir``; returns both paths."""
    outdir.mkdir(parents=True, exist_ok=True)

    rows: list[dict] = []
    classes: dict[str, DiffGeometry | None] = {}
    pair_groups_seen: set[frozenset[str]] = set()

    for sc in sheets:
        c = sc.circuit
        for net in c.nets.values():
            if net.net_class != NetClass.PORT:
                continue
            pt = c.port_type_of(net.name)
            ncls = _net_class(pt.kind, pt.impedance, pt.level_v)
            geo = GEOMETRY.get(pt.impedance) if pt.impedance else None
            classes.setdefault(ncls, geo)
            group, tol = "", ""
            if pt.pair_with:
                key = frozenset((net.name, pt.pair_with))
                pair_groups_seen.add(key)
                group = "PAIR:" + "/".join(sorted(key))
                tol = f"{INTRA_PAIR_SKEW_MM[pt.kind]:g}"
            elif pt.kind == "sd_bus":
                group = f"BUS:{pt.bus or 'SD'}"
                tol = f"{SD_BUS_MATCH_MM:g}"
            rows.append({
                "net": net.name,
                "sheet": sc.name,
                "kind": pt.kind,
                "net_class": ncls,
                "impedance_ohm": pt.impedance or "",
                "track_width_mm": f"{geo.width_mm:g}" if geo else "",
                "pair_gap_mm": f"{geo.gap_mm:g}" if geo else "",
                "pair_with": pt.pair_with or "",
                "length_match_group": group,
                "match_tolerance_mm": tol,
                "notes": "; ".join(filter(None, (
                    f"bus={pt.bus}" if pt.bus else "",
                    f"speed={pt.speed_hz}Hz" if pt.speed_hz else "",
                    f"level={pt.level_v:g}V" if pt.level_v is not None else "",
                    f"deferred: {pt.expect}" if pt.expect else "",
                    geo.source if geo else ""))),
            })

    # ---- .kicad_dru -----------------------------------------------------------
    dru_lines = [
        "(version 1)",
        "",
        "# Generated by schgen/constraints.py from typed ports.",
        "# Stackup: JLCPCB JLC04161H-7628 (4L 1.6mm, 7628 prepreg 0.2104mm).",
        "# Geometry source: JLCPCB impedance calculator for this stackup",
        "# (90R diff: 0.2611/0.2032mm; 100R diff: 0.2052/0.2032mm, outer vs L2).",
        "# Assign nets to the classes listed in layout_constraints.csv first.",
        "",
    ]
    for ncls, geo in sorted(classes.items()):
        if geo is None:
            continue
        dru_lines += [
            f'(rule "{ncls}_geometry"',
            f'  (condition "A.NetClass == \'{ncls}\'")',
            f"  (constraint track_width (min {geo.width_mm}mm) "
            f"(opt {geo.width_mm}mm) (max {geo.width_mm}mm))",
            f"  (constraint diff_pair_gap (min {geo.gap_mm}mm) "
            f"(opt {geo.gap_mm}mm))",
            ")",
            "",
        ]
    dru_path = outdir / "layout_constraints.kicad_dru"
    dru_path.write_text("\n".join(dru_lines))

    # ---- CSV ------------------------------------------------------------------
    csv_path = outdir / "layout_constraints.csv"
    fieldnames = ["net", "sheet", "kind", "net_class", "impedance_ohm",
                  "track_width_mm", "pair_gap_mm", "pair_with",
                  "length_match_group", "match_tolerance_mm", "notes"]
    with open(csv_path, "w", newline="") as f:
        w = csv.DictWriter(f, fieldnames=fieldnames)
        w.writeheader()
        for row in sorted(rows, key=lambda r: (r["net_class"], r["sheet"],
                                               r["net"])):
            w.writerow(row)
    return dru_path, csv_path

File: native/src/constraints.cpp
#include "schgen/constraints.hpp"
#include "schgen/occupancy.hpp"
#include "schgen/atomic_file.hpp"

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <limits>
#include <locale>
#include <map>
#include <set>
#include <sstream>
#include <tuple>

namespace schgen {
namespace {
const JsonNode& get(const JsonNode& n, const std::string& key, JsonKind kind) {
    const auto* p = object_field(n, key);
    if (!p || p->kind != kind) throw ProjectError("SI specification requires " + key + " with the correct type");
    return *p;
}
std::string optional_text(const JsonNode& n, const std::string& key) {
    return object_field(n, key) ? get(n, key, JsonKind::String).string_value : "";
}
std::string general(double value) {
    if (!std::isfinite(value)) throw ProjectError("non-finite layout constraint");
    std::ostringstream stream;
    stream.imbue(std::locale::classic()); stream << std::setprecision(6) << value;
    return stream.str();
}
std::string quote_csv(const std::string& value) {
    if (value.find_first_of(",\"\r\n") == std::string::npos) return value;
    std::string result = "\"";
    for (char c : value) { if (c == '"') result += '"'; result += c; }
    return result + '"';
}
std::string join(const std::vector<std::string>& items, const std::string& separator) {
    std::string result;
    for (const auto& item : items) { if (!result.empty()) result += separator; result += item; }
    return result;
}
}
double PairSignalSpec::match_tol_mm() const { return py_round(match_tol_mil * 0.0254, 4); }
double PairSignalSpec::intra_pair_skew_mm() const { return py_round(intra_pair_skew_mil * 0.0254, 4); }
std::vector<PairSignalSpec> parse_signal_specs(const JsonNode& data) {
    if (data.kind != JsonKind::Object) throw ProjectError("SI specification must be an object");
    const auto* pairs = object_field(data, "pairs");
    if (!pairs) return {};
    if (pairs->kind != JsonKind::Array) throw ProjectError("SI pairs must be an array");
    std::vector<PairSignalSpec> result;
    for (const auto& p : pairs->array_value) {
        PairSignalSpec spec;
        spec.interface = get(p,"interface",JsonKind::String).string_value;
        spec.signal = optional_text(p,"signal");
        spec.net_p = get(p,"net_p",JsonKind::String).string_value;
        spec.net_n = get(p,"net_n",JsonKind::String).string_value;
        const double impedance = get(p,"z_diff_ohm",JsonKind::Number).number_value;
        if (!std::isfinite(impedance) || impedance < std::numeric_limits<int>::min() || impedance > std::numeric_limits<int>::max())
            throw ProjectError("SI impedance out of range");
        spec.z_diff_ohm = static_cast<int>(impedance);
        spec.match_tol_mil = get(p,"match_tol_mil",JsonKind::Number).number_value;
        spec.intra_pair_skew_mil = get(p,"intra_pair_skew_mil",JsonKind::Number).number_value;
        if (!std::isfinite(spec.match_tol_mil) || !std::isfinite(spec.intra_pair_skew_mil)) throw ProjectError("non-finite SI target");
        spec.ac_coupled = get(p,"ac_coupled",JsonKind::Bool).bool_value;
        spec.spec_cite = get(p,"spec_cite",JsonKind::String).string_value;
        spec.notes = optional_text(p,"notes");
        result.push_back(std::move(spec));
    }
    std::stable_sort(result.begin(),result.end(),[](const auto& a,const auto& b){
        return std::tie(a.interface,a.net_p) < std::tie(b.interface,b.net_p);
    });
    return result;
}
std::vector<PairSignalSpec> load_signal_specs(const std::filesystem::path& path) {
    return parse_signal_specs(parse_json_file(path.string()));
}
const DifferentialGeometry* differential_geometry(int impedance) {
    static const DifferentialGeometry usb{90,0.2611,0.2032,"JLCPCB calculator, JLC04161H-7628 outer/L2"};
    static const DifferentialGeometry tmds{100,0.2052,0.2032,"JLCPCB calculator, JLC04161H-7628 outer/L2"};
    return impedance == 90 ? &usb : impedance == 100 ? &tmds : nullptr;
}
std::string port_net_class(const CircuitPortIr& p) {
    if (p.kind == "usb_hs_pair") return "DP90_USB";
    if (p.kind == "tmds_pair") return "DP100_TMDS";
    if (p.kind == "diff_pair") return "DP" + (p.has_impedance ? std::to_string(p.impedance) : "None") + "_DIFF";
    if (p.kind == "i2c") return "I2C";
    if (p.kind == "sd_bus") {
        if (!p.has_level_v) throw ProjectError("SD bus needs a voltage level");
        auto level = general(p.level_v); std::replace(level.begin(),level.end(),'.','V');
        return "SD_" + level;
    }
    return "Default";
}
bool needs_signal_specs(const std::vector<ProjectCircuit>& sheets) {
    for (const auto& sheet : sheets) for (const auto& net : sheet.circuit.nets)
        if (net.net_class == "port" && !circuit_port_type(sheet.circuit,net.name).pair_with.empty()) return true;
    return false;
}
LayoutConstraints generate_layout_constraints(const std::vector<ProjectCircuit>& sheets,
        const std::vector<PairSignalSpec>& specs, const std::string& source) {
    std::map<std::string,const PairSignalSpec*> by_net;
    for (const auto& spec : specs) { by_net[spec.net_p] = &spec; by_net[spec.net_n] = &spec; }
    std::map<std::string,const DifferentialGeometry*> classes;
    std::vector<std::vector<std::string>> rows;
    for (const auto& sheet : sheets) for (const auto& net : sheet.circuit.nets) {
        if (net.net_class != "port") continue;
        const auto p = circuit_port_type(sheet.circuit,net.name);
        const auto cls = port_net_class(p);
        const auto* geometry = p.has_impedance ? differential_geometry(p.impedance) : nullptr;
        classes.emplace(cls,geometry);
        std::string group,tolerance,skew_cite;
        if (!p.pair_with.empty()) {
            const auto found = by_net.find(net.name);
            if (found == by_net.end()) throw ProjectError(net.name + ": typed " + p.kind + " but has no row in " + source +
                " — every pair's intra-pair skew is read per net from the researched table; there is no per-kind default to fall back to. Add the pair with its standard citation.");
            std::set<std::string> pair{net.name,p.pair_with};
            group = "PAIR:" + join(std::vector<std::string>(pair.begin(),pair.end()),"/");
            tolerance = general(found->second->intra_pair_skew_mm());
            skew_cite = "skew: " + found->second->spec_cite;
        } else if (p.kind == "sd_bus") { group = "BUS:" + (p.bus.empty() ? "SD" : p.bus); tolerance = "2.5"; }
        std::vector<std::string> notes;
        if (!p.bus.empty()) notes.push_back("bus="+p.bus);
        if (p.has_speed_hz && p.speed_hz) notes.push_back("speed="+std::to_string(p.speed_hz)+"Hz");
        if (p.has_level_v) notes.push_back("level="+general(p.level_v)+"V");
        if (!p.expect.empty()) notes.push_back("deferred: "+p.expect);
        if (geometry) notes.push_back(geometry->source);
        if (!skew_cite.empty()) notes.push_back(skew_cite);
        rows.push_back({net.name,sheet.name,p.kind,cls,p.has_impedance && p.impedance ? std::to_string(p.impedance) : "",
            geometry ? general(geometry->width_mm) : "",geometry ? general(geometry->gap_mm) : "",p.pair_with,
            group,tolerance,join(notes,"; ")});
    }
    std::stable_sort(rows.begin(),rows.end(),[](const auto& a,const auto& b){
        return std::tie(a[3],a[1],a[0]) < std::tie(b[3],b[1],b[0]);
    });
    LayoutConstraints result; result.port_count = rows.size();
    result.csv = "net,sheet,kind,net_class,impedance_ohm,track_width_mm,pair_gap_mm,pair_with,length_match_group,match_tolerance_mm,notes\r\n";
    for (const auto& row : rows) {
        for (std::size_t i = 0; i < row.size(); ++i) { if (i) result.csv += ','; result.csv += quote_csv(row[i]); }
        result.csv += "\r\n";
    }
    result.dru = "(version 1)\n\n# Generated by schgen/constraints.py from typed ports.\n"
        "# Stackup: JLCPCB JLC04161H-7628 (4L 1.6mm, 7628 prepreg 0.2104mm).\n"
        "# Geometry source: JLCPCB impedance calculator for this stackup\n"
        "# (90R diff: 0.2611/0.2032mm; 100R diff: 0.2052/0.2032mm, outer vs L2).\n"
        "# Assign nets to the classes listed in layout_constraints.csv first.\n";
    for (const auto& [name,geo] : classes) {
        if (!geo) continue;
        const auto width = general(geo->width_mm), gap = general(geo->gap_mm);
        result.dru += "\n(rule \"" + name + "_geometry\"\n  (condition \"A.NetClass == '" + name + "'\")\n"
            "  (constraint track_width (min " + width + "mm) (opt " + width + "mm) (max " + width + "mm))\n"
            "  (constraint diff_pair_gap (min " + gap + "mm) (opt " + gap + "mm))\n)\n";
    }
    return result;
}
LayoutConstraints write_layout_constraints(const std::vector<ProjectCircuit>& sheets,
        const std::filesystem::path& specification, const std::filesystem::path& directory) {
    const auto specs = needs_signal_specs(sheets) ? load_signal_specs(specification) : std::vector<PairSignalSpec>{};
    auto result = generate_layout_constraints(sheets,specs,specification.string());
    write_atomic_file((directory/"layout_constraints.kicad_dru").string(),{result.dru.begin(),result.dru.end()});
    write_atomic_file((directory/"layout_constraints.csv").string(),{result.csv.begin(),result.csv.end()});
    return result;
}
}  // namespace schgen
